drivers/sensor/grow_r502a/grow_r502a.c

```diff
@@ -436,152 +650,323 @@ static int fps_empty_db(const struct device *dev)
+	char const upload_temp_len = 2;
+	int ret = 0, idx = 0;
+
+	if (!temp->data) {
+		LOG_ERR("Invalid temp data");
+		return -EINVAL;
+	}
+
+	union r502a_packet tx_packet = {
+		.pid = R502A_COMMAND_PACKET,
+		.data = {R502A_UPCHAR, R502A_CHAR_BUF_1}
+	};
 
 	k_mutex_lock(&drv_data->lock, K_FOREVER);
 
-	ret = fps_delete_model(dev, val->val1, val->val2);
+	ret = transceive_packet(dev, &tx_packet, &rx_packet, upload_temp_len);
 	if (ret != 0) {
 		goto unlock;
 	}
 
-	ret = fps_get_template_count(dev);
+	ret = r502a_validate_rx_packet(&rx_packet);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	if (rx_packet.buf[R502A_CC_IDX] == R502A_OK) {
+		LOG_DBG("Upload to host controller");
+	} else {
+		LOG_ERR("Error uploading template 0x%X",
+					rx_packet.buf[R502A_CC_IDX]);
+		ret = -EIO;
+		goto unlock;
+	}
+
+	do {
+		ret = transceive_packet(dev, NULL, &rx_packet, 0);
+		if (ret != 0) {
+			goto unlock;
+		}
+
+		ret = r502a_validate_rx_packet(&rx_packet);
+		if (ret != 0) {
+			goto unlock;
+		}
+
+		memcpy(&temp->data[idx], &rx_packet.data,
+				sys_be16_to_cpu(rx_packet.len) - R502A_CHECKSUM_LEN);
+		idx += sys_be16_to_cpu(rx_packet.len) - R502A_CHECKSUM_LEN;
+	} while (rx_packet.pid != R502A_END_DATA_PACKET);
+
+	temp->len = idx;
 
 unlock:
 	k_mutex_unlock(&drv_data->lock);
 	return ret;
 }
 
-static int fps_match(const struct device *dev, struct sensor_value *val)
+/**
+ * @brief	download template from controller to sensor device's RAM buffer.
+ */
+static int fps_download_char_buf(const struct device *dev, uint8_t char_buf_id,
+						const struct sensor_value_ex *temp)
 {
 	struct grow_r502a_data *drv_data = dev->data;
-	int ret = -1;
+	union r502a_packet rx_packet = {0};
+	char const down_temp_len = 2;
+	int ret = 0, i = 0;
+
+	char_buf_id = (char_buf_id > R502A_CHAR_BUF_2) ? R502A_CHAR_BUF_2 : R502A_CHAR_BUF_1;
+
+	if (!temp->data || (temp->len < R502A_TEMPLATE_MAX_SIZE)) {
+		LOG_ERR("Invalid temp data");
+		return -EINVAL;
+	}
+
+	union r502a_packet tx_packet = {
+		.pid = R502A_COMMAND_PACKET,
+		.data = {R502A_DOWNCHAR, char_buf_id}
+	};
 
 	k_mutex_lock(&drv_data->lock, K_FOREVER);
 
-	ret = fps_get_image(dev);
+	ret = transceive_packet(dev, &tx_packet, &rx_packet, down_temp_len);
 	if (ret != 0) {
 		goto unlock;
 	}
 
-	ret = fps_image_to_char(dev, R502A_CHAR_BUF_1);
+	ret = r502a_validate_rx_packet(&rx_packet);
 	if (ret != 0) {
 		goto unlock;
 	}
 
-	ret = fps_search(dev, R502A_CHAR_BUF_1);
-	if (ret == 0) {
-		val->val1 = drv_data->finger_id;
-		val->val2 = drv_data->matching_score;
+	if (rx_packet.buf[R502A_CC_IDX] == R502A_OK) {
+		LOG_DBG("Download to R502A sensor");
+	} else {
+		LOG_ERR("Error downloading template 0x%X",
+					rx_packet.buf[R502A_CC_IDX]);
+		ret = -EIO;
+		goto unlock;
+	}
+
+	while (i < (R502A_TEMPLATE_MAX_SIZE - R502A_DATA_PKT_SIZE)) {
+		tx_packet.pid = R502A_DATA_PACKET;
+		memcpy(tx_packet.data, &temp->data[i], R502A_DATA_PKT_SIZE);
+
+		ret = transceive_packet(dev, &tx_packet, NULL, R502A_DATA_PKT_SIZE);
```

Comment:
does R502A_DATA_PKT_SIZE magic 128 doesn't affects if sys parameter changes the data packet size?

Reply:
yes it will, is it better to have Kconfig for this with default point to 128, if needed user can change it to 32 or 64 or 128 or 256

Reply:
addressed in the latest changes.